### ISSUE TITLE
Error Formatting

### ISSUE THREAD

**Issue by Eluinhost** [dchester/epilogue#45]
## Defaults formatting
Errors should be in the format:

```json
{
  "message": "error message",
  "errors": ["array", "of", "problems"]
}
```
The following show just 'message' when returning an error:

https://github.com/dchester/epilogue/blob/master/lib/Controllers/read.js#L45
https://github.com/dchester/epilogue/blob/master/lib/Controllers/delete.js#L30

I don't think these are a problem, the errors array should be optional.

This file is just showing `error` for a 404 though:
https://github.com/dchester/epilogue/blob/master/lib/Controllers/update.js#L33

Plus this line just returns `error` for a 500:
https://github.com/dchester/epilogue/blob/master/lib/Controllers/list.js#L120

These should probably use a similar system to stop having to worry about fitting the same format each time.

## Returning Errors
Returning an error is awkward from milestones too, for example I have this middleware:

```javascript
module.exports = function(attributes) {
  if(!_.isArray(attributes)) attributes = [attributes];

  return function(req, res, context) {
    var intersect = _.intersection(_.keys(req.body), attributes);
    if(intersect.length) {
      res.status(400);
      context.error = {
        message: 'Cannot set fields on this object',
        errors: intersect
      };
      // skip the actual write
      return context.skip();
    }

    return context.continue();
  };
};
```

Here I am using the same system used in the default middlewares of doing:

```javascript
res.staus(400); //set error code
context.error = {
  message: 'error message',
  errors: ['array of problems']
}
context.skip(); // skip to next
```
The problem I have with this is it is now required to be ran on write.before so the skip works and then the write function needs to also check there is no errors to `context.continue`. On `create` the default write function doesn't check context.error but the one in `update` does. Every middleware called after this one needs to be aware of the error to avoid processing when there is an error and allow the error to pass through to the deafult .send action (on this note the complete callback says `Run the specified function when the request is complete, regardless of the status of the response.` but as far as I can tell if context.stop() is called it won't run)

# Potential Solution
I think a proper way to handle this would be to add error to context as a function.

```javascript
context.error({
  message: 'error message',
  status: 400,  // optional and default 500?
  errors: ['array of problems'] // optional
});
```
Calling error should then call a special 'error' function that is outside of the chain (and probably no .before or .after) where the default can be:

```javascript
function(req, res, context, error) {
  res.status(error.status);
  res.json(_.omit(error, ['status']));
  context.stop();
}
```

This would make keeping to the default format within epilogue simple as they would just need to call `context.error(...)` and would no longer need to check `context.error` and continue if it is set. It would also allow customization of the default middleware functions error rendering.

It should also allow context.continue() and context.skip() to resume processing from within the error milestone.

It could also be tier structured:

Running create.write.after would check for and run the first of:

```
model.create.write.after.error  
model.create.write.error  
model.create.error  
model.error  
default error function  
```

And update.fetch would check for

```
model.update.fetch.action.error
model.update.fetch.error
model.update.error
model.error
default error function
```

If the tiering isn't possible at minimum a 'previous state' variable should be accessible in the error function to tell what hook errored and allow separate processing based on the state

On another side I also see this:
https://github.com/dchester/epilogue/blob/master/lib/Controllers/base.js#L122

Which should catch any uncaught errors, but just passes them to an empty function

**Comment by mbroadst** [comment]
context.error is certainly one way to handle it, but another alternative would be to convert the "work" step (in [Controller.prototype._control](https://github.com/dchester/epilogue/blob/master/lib/Controllers/base.js#L86)) to use promises, rather than async. I have a PoC showing off something similar for sequelize hooks [here](https://github.com/mbroadst/sequelize-hooks-wip).

Using this approach we could actually reduce the error processing/serialization to one location, and instead of explicitly calling context.error() (or setting the context.error variable and calling context.stop), one could simply ```throw new EpilogueError("some error", ["some", "errors"])``` *or* just return the promise from Sequelize. Thoughts?

**Comment by Eluinhost** [comment]
Yeah being able to throw on errors would be the best (and most natural) way of doing it. It would let us customize the error render based on the error class being thrown. I would love to be able to just throw an InvalidParametersError and then have it automatically render a 400 with a list of the invalid fields

**Comment by Eluinhost** [comment]
@mbroadst today I took a go at converting to promises in the hooks and fixing the issues with error handling in [this branch](https://github.com/Eluinhost/epilogue/tree/features/promises). 

All the tests are in a passing state, there is still a little bit to go though. On a side note on my work PC I am getting intermittent errors with a few different tests (like 'should support a generic query string as well as other criteria'), but I dont remember ever having them fail at home, I'm going to assume its the machine setup :)

# Changes

## context.continue() and context.skip()

The milestone function now expects a truthy/falsey value to be returned or a promise that resolves to one.

If the value is falsey it is the equivalent of `context.continue()` and continues execution. This allows a function to return nothing and it will still continue to execute hooks

If the value is truthy it is the equivalent of `context.skip()` and will continue to execute other functions on the same hook but will skip the next hook in the series. e.g. if create.before[0] returns true then create.before[1] would be called but all of create.action[] would be skipped.
*I couldn't quite work out if this was how it was running originally or not, please tell me if I got this wrong*

## context.stop()

Removed in favour of throwing an error. If data has already been written to the client a special 'error' `RequestCompleted` can be thrown that has no additional writing.

*I don't like using an error when there isn't an actual error, how else can this be done? We could return/resolve to `'done'` and put a special check in at each milestone, but I still think an error would need to be thrown to break the promise chain in the background*


## Errors

Errors are now displayed by throwing an error in the milestone function. This stops execution of the hook and any future hooks and will render the error to the client.

Errors that extend EpilogueError will display in the format:

```json
{
  "message": err.message,
  "errors": err.errors
}
```

and will have the HTTP status code of `err.statusCode`.

There are a few predefined errors:

`BadRequestError` - forces status code 400
`ForbiddenError` - forces status code 403
`NotFoundError` - forces status code 404
`RequestCompleted` - shows no additional output, special case

Errors that do not extend EpilogueError will be display as:

```json
{
  "message": "internal error",
  "errors": [ err.message ]
}
```

Sequelize validation errors are automatically caught and displayed the same as was previously handled

*There should be a way to register an error handlers for a particular error classes and override the default handling too, not too sure about the best way to do this*

I have just commented out the tests of the type 'should support modifying error data on create before sending response' as I think they are not needed anymore but I didnt want to delete them just yet.

**Comment by mbroadst** [comment]
@Eluinhost ah! I had a bunch of this done on a local branch but it looks like you beat me to it :smile:. Generally this aligns with my approach, but I'm a little concerned of breaking the context.continue/skip/stop behavior for existing users. With my maintainer hat on this is kind of a big no-no, have you played around with ways to preserve that behavior? (for instance, throwing RequestCompleted on context.stop()).

I generally feel that context.continue() can go away, as it literally does nothing, but I could see how it helps readability for some users. Also, it seems that your context.skip() behavior changes existing behavior? Specifically, in the current codebase context.skip() skips the rest of the current hooks as well as the next milestone.

I'll give your branch a build later tonight and see if I'm running into the same issues you are with flakey tests, but generally I have not previously had issues. Are you experiencing those same issues with the async version of this code?

**Comment by mbroadst** [comment]
@Eluinhost also from a purely aesthetic (nitpicky) standpoint I'd say you could consolidate all the errors to just lib/errors.js, rather than a separate file for each of them. As well as EpilogueError should take the signature (status, message, errors), status being the most important element in a REST api.

I was also playing around with more comprehensive stack traces for errors, mentioned at the end of this [section](https://github.com/petkaantonov/bluebird/blob/master/API.md#catchfunction-errorclassfunction-predicate-function-handler---promise) of bluebird's documentation. I'm not sure what sort of performance penalty we would pay for that, but I imagine it's negligible when taking into account the usefulness of a proper stack trace.

**Comment by Eluinhost** [comment]
@mbroadst

For context.continue we can just use a noop and things will have the same effect. For stop we can also throw a RequestCompleted error. Not sure how we can make skip work the same here apart from doing the same with a MilestoneSkip type error that can be caught per hook and then cause the hook to resolve to true (and put deprecation notices on all of these?)

I'll have to switch the skip function logic, like I said in the post above I wasn't quite sure how the .skip() worked currently so I made it complete the current set of hooks before skipping the next milestone. 

As for the failed tests the same thing was happening on the async version, it ways always on tests that deep equals results and they were different elements. Tomorrow I'll take a look at my node version and see if I can reproduce the same errors on another machine, that machine also seems to have problems with _.isInteger too for some reason.

I've pushed the changes to errors, I agree status should come first. I've also made the RequestCompleted error not part of the EpilogueError inheritance tree to stop it requiring the stack trace e.t.c.

**Comment by mbroadst** [comment]
@Eluinhost it actually looks from a quick glance at your changes that return a truthy value skips right? So why not just have ```context.skip = function() { return true; }``` and ```return context.skip()``` skips the milestone?

**Comment by Eluinhost** [comment]
@mbroadst the problem with that is that would break backwards compatability for people that just called `context.skip()` without returning it or called async code and inside callbacks doing `context.skip()`

**Comment by mbroadst** [comment]
@Eluinhost ah yes, good point. Hrm, sounds like the only way to do it is to have an extra SkipMilestone error I guess but that feels like a misuse of throwing errors :smile:

**Comment by Eluinhost** [comment]
@mbroadst You said `Specifically, in the current codebase context.skip() skips the rest of the current hooks as well as the next milestone.` but the main documentation says `Skip to the next milestone, usually called from before.`. I feel like it might just be a terminology thing causing my confusion so I'm calling them these:

- Milestone - auth/fetch e.t.c.
- Hook - auth.before[] auth.action[] and auth.after[]
- Hook Function - auth.before[0] e.t.c.

I had to do this to avoid confusing myself by calling each function a hook when the entire _before was technically. 

Looking at the main documentation it looks to me like when a Hook Function wants to skip then it should skip execution of every remaining hook function in the milestone. So if `auth.before[1]` asks to skip then only `auth.before[0]` and `auth.before[1]` will run and `auth.before[2:]`, `auth.action[:]` and `auth.after[:]` will be skipped and execution goes back to normal for `fetch.before[0]`, so I've implemented that way of working and added a couple of tests for it.

I've taken a look at keeping support for `context.skip()` but even throwing an error is not going to work for it. For example:

```javascript
auth.before(function(req, res, context) {
  doSomeAsyncStuff(function(authed) {
    if(authed) {
      context.skip();
    } else {
      context.continue();
    }
  })
});
```

This would appear to work for `context.continue` calls but in reality the function returns `undefined` and so will continue no matter what, the `context.skip` call would come too late. This also means calling context.stop() wouldn't work inside their async code.

The alternative here is to check the return value of every hook for undefined and if it is undefined then rely on a context object to move to the next promise:

```javascript
hookFunctionChain = hookFunctionChain.then(function runHookFunction(skipNext) {
    // if any asked to skip keep returning true to avoid calling further functions inside this hook
    if(skipNext) return true;

    var contextPromise = new Promise();
    context.skip = function() {
        contextPromise.resolve('skip');
    };
    context.stop = function() {
        contextPromise.resolve('stop');  
    };
    context.continue = function() {
        contextPromise.resolve('continue');
    };

    Promise.resolve(f.call(self, req, res, context))
        .then(function(skip) {
            if(!_.isUndefined(skip)) {
                return skip;
            }
            
            contextPromise.then(function(whatdo) {
                if(whatdo === 'stop') throw new Errors.RequestCompleted;
                
                return whatdo === 'skip';
            });          
        });
});
```

This would require a hook function to return/resolve true/false otherwise it would check for context.x() calls instead. This way would keep the existing syntax working but would still have a few issues if people are using `return` calls in their current function with things that are defined and error handling is different too so things will need to be reworked anyway.

If we are going to use a different error system and promises then in my opinion this will have to be breaking changes with a version bump and just drop support for the context.x() functions.

**Comment by mbroadst** [comment]
@Eluinhost your understanding of the order of operations for the hooks is correct, if you wanted to amend the documentation to include that description it would be much appreciated (clearly its a point of confusion). 

**context.skip():** I see what you're saying, but I wonder if your experience would change if we altered your first example above to be more like:

```
auth.before(function(req, res, context) {
  return doSomeAsyncStuff(function(authed) {
    if(authed) {
      return context.skip();
    } else {
      return context.continue();
    }
  })
});
```

Again, I'm not really against making a breaking change for technical issues we can't overcome, but I think the pill is easier to swallow if we make an attempt to maintain the existing mental model (so you still get to use context.continue/skip/stop, but now you need to remember that you must *return* those calls, rather than just making them, for instance). Thoughts?

**Comment by Eluinhost** [comment]
@mbroadst I've tried a few different things out today and what I've got in the branch right now is what I think is a better approach. 

It allows old style callbacks:

```javascript
auth.before(function(req, res, context) {
  doSomeAsyncStuff(function(err, authed) {
    if (authed) {
      context.skip();
    } else {
      context.continue();
    }
  })
});
```

It allows sync returns:

```javascript
auth.before(function(req, res, context) {
  if(authed) {
    return context.skip;
  } else {
    return context.continue;
  }
});
```

And it allows for promise resolving:

```javascript
auth.before(function(req, res, context) {
  return doSomeAsyncStuff().then(function(authed) {
    if (authed) {
      return context.skip;
    } else {
      return context.continue;
    }
  })
});
```

It also allows for context.stop for use in the same way so we get rid of public use of RequestCompletedError.

Basically if the function does this if it returns/resolve to them:

- context.error - it throws a RequestCompletedError to break the chain silently
- context.continue - returns false for the skip
- context.skip - returns true for the skip
- anything else - waits on a secondary decision promise

The context.error(), context.skip() and context.continue() functions all resolve a secondary decision promise in the background that is only waited for if the function didn't return/resolve to a valid value.

This allows any existing methods to return anything they like and still use context.continue() e.t.c. as if nothing has changed but still allow returning a promise/value in newer versions.

I've still got to write some test cases to verify all of the use cases work as expected, but I think this way is a lot clearer in the user code about what is happening compared to `return true` and `return false`

**Comment by mbroadst** [comment]
@Eluinhost sounds fantastic. I'd say at this point when you feel like you've tested it enough just go ahead and put up a PR

**Comment by mterwill** [comment]
Question: if we are defining a milestone, the documentation makes it appear that we could just throw an error as-is. This does not work for me. Is there a file I should be including, or something else I may be missing?

Here's the code from README

````
users.delete.auth(function(req, res, context) {
    throw new ForbiddenError("can't delete a user");
})
````

and a snippet of my code

````
room_tasks.list.fetch.before(function(req, res, context) {
    var room = req.params['room'];
    if(isValidRoom(room)) {
        context.criteria = { room : req.params['room'] };
        return context.continue();
    }
    
    throw new NotFoundError();
});
````

which yields a response ``500``
````
{
    message: "internal error",
    errors: [
        "NotFoundError is not defined"
    ]
}
````

Thanks for all your work on this! Has been a pleasure to use and very effective in streamlining my API implementation.

**Comment by mbroadst** [comment]
@mterwill to use NotFoundError you'll need to do ```var NotFoundError = require('epilogue').Errors.NotFoundError``. Optionally you can also just use context.error(status, message, errors)

**Comment by mterwill** [comment]
Thanks!

**Comment by mbroadst** [comment]
ya, I'll go ahead and update those docs :smile:

**Comment by mbroadst** [comment]
@mterwill added with da58c8b7ec1f4243e3bda02489c71d23a926b265
